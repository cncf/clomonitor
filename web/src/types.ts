export interface Organization {
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
}

export interface BaseProject {
  id: string;
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
  devstats_url?: string;
  maturity_id: Maturity;
  category_id: Category;
  score: Score;
  updated_at: number;
}

export interface Project extends BaseProject {
  repositories: BaseRepository[];
  organization: {
    name: string;
  };
}

export interface ProjectDetail extends BaseProject {
  repositories: Repository[];
}

export interface BaseRepository {
  name: string;
  url: string;
  kind: RepositoryKind;
}

export interface Repository extends BaseRepository {
  digest: string;
  repository_id: string;
  score: Score;
  reports: Report[];
}

export interface Report {
  data: CoreReport | any;
  linter_id: LinterId;
  report_id: string;
  updated_at: number;
}

export interface CoreReport {
  // ScoreType
  [key: string]: {
    [key: string]: ReportCheck;
  };
}

export interface ReportCheck {
  passed: boolean;
  value?: string;
  url?: string;
}

export type Score = {
  [key in ScoreType]: number;
} & { score_kind: ScoreKind };

export interface FiltersSection {
  name: string;
  title: string;
  filters: Filter[];
}

export interface Filter {
  name: string | number;
  label: string;
  legend?: string;
  decorator?: JSX.Element;
}

export interface Issue {
  level: number;
  description: string;
}

export interface Prefs {
  search: { limit: number; sort: { by: SortBy; direction: SortDirection } };
  theme: {
    effective: string;
  };
}

export interface ReportOptionData {
  icon: JSX.Element;
  name: string;
  legend: JSX.Element;
}

export enum Maturity {
  Graduated = 0,
  Incubating,
  Sandbox,
}

export enum Category {
  'App definition' = 0,
  Observability,
  Orchestration,
  Platform,
  Provisioning,
  Runtime,
  Serverless,
}

export enum Rating {
  A = 'a',
  B = 'b',
  C = 'c',
  D = 'd',
}

export enum FilterKind {
  Maturity = 'maturity',
  Category = 'category',
  Rating = 'rating',
}

export enum LinterId {
  core = 0,
}

export enum ScoreType {
  BestPractices = 'best_practices',
  Documentation = 'documentation',
  Global = 'global',
  Legal = 'legal',
  License = 'license',
  Security = 'security',
}

export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortBy {
  Name = 'name',
  Score = 'score',
}

export enum RepositoryKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ScoreKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ReportOption {
  Adopters = 'adopters',
  ApprovedLicense = 'approved',
  ArtifactHubBadge = 'artifacthub_badge',
  Changelog = 'changelog',
  CodeOfConduct = 'code_of_conduct',
  CommunityMeeting = 'community_meeting',
  Contributing = 'contributing',
  DCO = 'dco',
  Governance = 'governance',
  LicenseScanning = 'scanning',
  Maintainers = 'maintainers',
  OpenSSFBadge = 'openssf_badge',
  Readme = 'readme',
  RecentRelease = 'recent_release',
  Roadmap = 'roadmap',
  SecurityPolicy = 'security_policy',
  SlackPresence = 'slack_presence',
  SPDX = 'spdx_id',
  TrademarkDisclaimer = 'trademark_disclaimer',
  Website = 'website',
}

export interface SearchFiltersURL extends BasicQuery {
  pageNumber: number;
}

export interface BasicQuery {
  text?: string;
  filters?: {
    [key: string]: (string | number)[];
  };
}

export interface SearchQuery extends BasicQuery {
  limit: number;
  offset: number;
  sort_by: SortBy;
  sort_direction: SortDirection;
}

export interface SearchData {
  limit: number;
  offset: number;
  sort_by: string;
  sort_direction: string;
  text?: string;
  category?: number[];
  maturity?: number[];
  rating?: number[];
}

export interface Error {
  kind: ErrorKind;
  message?: string;
}

export enum ErrorKind {
  Other,
  NotFound,
}

export type ReportOptionInfo = {
  [key in ReportOption]: ReportOptionData;
};

export interface RecommendedTemplate {
  name: string;
  url: string;
}
