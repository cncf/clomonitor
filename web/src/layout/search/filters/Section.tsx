import { isUndefined } from 'lodash';
import { ChangeEvent } from 'react';

import { Filter, FiltersSection } from '../../../types';
import CheckBox from '../../common/Checkbox';
import styles from './Section.module.css';

interface Props {
  section: FiltersSection;
  device: string;
  activeFilters?: string[];
  onChange: (name: string, value: string, checked: boolean) => void;
}

const Section = (props: Props) => {
  return (
    <>
      <div className={`fw-bold text-uppercase text-primary ${styles.categoryTitle}`}>
        <small>{props.section.title}</small>
      </div>

      <div className="mt-2">
        {props.section.filters.map((filter: Filter) => {
          return (
            <CheckBox
              key={`filter_${filter.name.toString()}`}
              name={props.section.name}
              value={filter.name.toString()}
              labelClassName="mw-100"
              legend={filter.legend}
              label={filter.label}
              icon={<span className={`position-relative ${styles.decorator}`}>{filter.decorator}</span>}
              device={props.device}
              checked={!isUndefined(props.activeFilters) && props.activeFilters.includes(filter.name.toString())}
              onChange={(e: ChangeEvent<HTMLInputElement>) =>
                props.onChange(e.target.name, e.target.value, e.target.checked)
              }
            />
          );
        })}
      </div>
    </>
  );
};

export default Section;
