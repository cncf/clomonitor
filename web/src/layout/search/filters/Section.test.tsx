import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { FilterKind, Rating } from '../../../types';
import Section from './Section';

const mockOnChange = jest.fn();

const defaultProps = {
  section: {
    name: FilterKind.Rating,
    title: 'Rating',
    filters: [
      {
        name: Rating.A,
        label: 'A',
        legend: '[75-100]',
      },
      {
        name: Rating.B,
        label: 'B',
        legend: '[50-74]',
      },
      {
        name: Rating.C,
        label: 'C',
        legend: '[25-49]',
      },
      {
        name: Rating.D,
        label: 'D',
        legend: '[0-24]',
      },
    ],
  },
  activeFilters: [],
  onChange: mockOnChange,
  device: 'test',
};

describe('Section', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<Section {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders Section', () => {
      render(<Section {...defaultProps} />);

      expect(screen.getByText('Rating')).toBeInTheDocument();
      expect(screen.getByRole('checkbox', { name: 'A [75-100]' })).toBeInTheDocument();
      expect(screen.getByRole('checkbox', { name: 'B [50-74]' })).toBeInTheDocument();
      expect(screen.getByRole('checkbox', { name: 'C [25-49]' })).toBeInTheDocument();
      expect(screen.getByRole('checkbox', { name: 'D [0-24]' })).toBeInTheDocument();
    });

    it('renders Section with selected options', () => {
      render(<Section {...defaultProps} activeFilters={['a', 'b']} />);

      expect(screen.getByRole('checkbox', { name: 'A [75-100]' })).toBeChecked();
      expect(screen.getByRole('checkbox', { name: 'B [50-74]' })).toBeChecked();
    });

    it('calls onChange to click filter', async () => {
      render(<Section {...defaultProps} />);

      const check = screen.getByRole('checkbox', { name: 'B [50-74]' });

      expect(check).not.toBeChecked();

      await userEvent.click(check);

      expect(mockOnChange).toHaveBeenCalledTimes(1);
      expect(mockOnChange).toHaveBeenCalledWith('rating', 'b', true);
    });

    it('calls onChange to click selected filter', async () => {
      render(<Section {...defaultProps} activeFilters={['b']} />);

      const check = screen.getByRole('checkbox', { name: 'B [50-74]' });

      expect(check).toBeChecked();

      await userEvent.click(check);

      expect(mockOnChange).toHaveBeenCalledTimes(1);
      expect(mockOnChange).toHaveBeenCalledWith('rating', 'b', false);
    });
  });
});
