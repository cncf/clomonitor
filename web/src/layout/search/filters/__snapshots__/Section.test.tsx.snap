// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="fw-bold text-uppercase text-primary categoryTitle"
  >
    <small>
      Rating
    </small>
  </div>
  <div
    class="mt-2"
  >
    <div
      class="form-check me-sm-2 mb-2 undefined"
    >
      <input
        aria-checked="false"
        class="form-check-input rounded-0 Checkbox-module_checkbox__lSeQj"
        id="test-rating-a"
        name="rating"
        tabindex="0"
        type="checkbox"
        value="a"
      />
      <label
        class="form-check-label mw-100"
        data-testid="checkboxLabel"
        for="test-rating-a"
      >
        <div
          class="d-flex align-items-baseline mw-100"
        >
          <span
            class="position-relative decorator"
          />
          <span
            class="text-truncate"
          >
            A
          </span>
          <small
            class="ps-1"
          >
            [75-100]
          </small>
        </div>
      </label>
    </div>
    <div
      class="form-check me-sm-2 mb-2 undefined"
    >
      <input
        aria-checked="false"
        class="form-check-input rounded-0 Checkbox-module_checkbox__lSeQj"
        id="test-rating-b"
        name="rating"
        tabindex="0"
        type="checkbox"
        value="b"
      />
      <label
        class="form-check-label mw-100"
        data-testid="checkboxLabel"
        for="test-rating-b"
      >
        <div
          class="d-flex align-items-baseline mw-100"
        >
          <span
            class="position-relative decorator"
          />
          <span
            class="text-truncate"
          >
            B
          </span>
          <small
            class="ps-1"
          >
            [50-74]
          </small>
        </div>
      </label>
    </div>
    <div
      class="form-check me-sm-2 mb-2 undefined"
    >
      <input
        aria-checked="false"
        class="form-check-input rounded-0 Checkbox-module_checkbox__lSeQj"
        id="test-rating-c"
        name="rating"
        tabindex="0"
        type="checkbox"
        value="c"
      />
      <label
        class="form-check-label mw-100"
        data-testid="checkboxLabel"
        for="test-rating-c"
      >
        <div
          class="d-flex align-items-baseline mw-100"
        >
          <span
            class="position-relative decorator"
          />
          <span
            class="text-truncate"
          >
            C
          </span>
          <small
            class="ps-1"
          >
            [25-49]
          </small>
        </div>
      </label>
    </div>
    <div
      class="form-check me-sm-2 mb-2 undefined"
    >
      <input
        aria-checked="false"
        class="form-check-input rounded-0 Checkbox-module_checkbox__lSeQj"
        id="test-rating-d"
        name="rating"
        tabindex="0"
        type="checkbox"
        value="d"
      />
      <label
        class="form-check-label mw-100"
        data-testid="checkboxLabel"
        for="test-rating-d"
      >
        <div
          class="d-flex align-items-baseline mw-100"
        >
          <span
            class="position-relative decorator"
          />
          <span
            class="text-truncate"
          >
            D
          </span>
          <small
            class="ps-1"
          >
            [0-24]
          </small>
        </div>
      </label>
    </div>
  </div>
</DocumentFragment>
`;
